File: src/main/java/team404/MaterialGenerator.java
package team404;

import lombok.Getter;
import org.apache.commons.lang3.tuple.Pair;
import org.bukkit.Material;
import team404.models.MaterialTier;
import team404.models.RequiredResource;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class MaterialGenerator {

    @Getter
    private static final Map<String, List<RequiredResource>> materialLists = new HashMap<>();

    public static void addMaterialList(String key, List<RequiredResource> materialList) {
        materialLists.put(key, materialList);
    }

    private static Pair<Integer, Material> getRandomPair(List<RequiredResource> list, MaterialTier tier) {
        Random rand = new Random();
        List<RequiredResource> filteredList = list.stream()
                .filter(m -> m.getTier().equals(tier))
                .toList();
        RequiredResource resource = filteredList.get(rand.nextInt(filteredList.size()));
        int lower = resource.getLowerThreshold();
        int higher = resource.getUpperThreshold();
        int count;
        if (higher == lower) {
            count = higher;
        } else {
            count = rand.nextInt(higher - lower) + lower;
        }
        return Pair.of(count, resource.getMaterial());
    }

    public static List<Pair<Integer, Material>> getMaterialsList(MaterialTier tier) {
        return materialLists.values().stream()
                .map(list -> getRandomPair(list, tier))
                .toList();
    }
}

File: src/main/kotlin/team404/service/MaterialGenerator.kt
package team404.service

import org.apache.commons.lang3.tuple.Pair
import org.bukkit.Material
import team404.model.MaterialTier
import team404.model.RequiredResource
import java.util.*

object MaterialGenerator {
    val materialLists: MutableMap<String, List<RequiredResource>> = HashMap()

    fun addMaterialList(key: String, materialList: List<RequiredResource>) {
        materialLists[key] = materialList
    }

    private fun getRandomPair(list: List<RequiredResource>, tier: MaterialTier): Pair<Int, Material> {
        val rand = Random()
        val filteredList = list.filter { it.tier == tier }
        return filteredList[rand.nextInt(filteredList.size)].let { (material, lower, higher, _) ->
            val count = if (higher == lower) {
                higher
            } else {
                rand.nextInt(higher - lower) + lower
            }
            Pair.of(count, material)
        }
    }

    fun getMaterialsList(tier: MaterialTier, multiplier: Int): Map<Material, Int> {
        val materialMap = HashMap<Material, Int>()
        for (i in 0 until multiplier) {
            materialLists.values
                .map { getRandomPair(it, tier) }
                .forEach { (amount, material) ->
                    materialMap[material] = materialMap.getOrDefault(material, 0) + amount
                }
        }
        return materialMap
    }
}
